### PR TITLE
Draw rectangular prisms for incomplete stabilizers 📦

### PR DESCRIPTION
**Pull request by sduquemesa**
## Context for changes

Whenever there are incomplete stabilizer, rectangular rather than cubic voxels are displayed.

- **Improvements**: 
    * Rectangular prisms drawn for incomplete stabilizers
    * Grey prisms for stabilizers with no bit values defined in the cluster state.
    * Grey edges on stabilizer graph when no bit values are defined.
 
## Example usage and tests

image

## Performance results justifying changes

- [x] Not Applicable

## Workflow actions and tests

- [x] Not Applicable

## Expected benefits and drawbacks

**Expected benefits:**
- Better visualization of stabilizers.

**Possible drawbacks:**
- None

## Related issues

Related Shortcut ticket [here](https://app.shortcut.com/xanaduai/story/21026/draw-rectangular-prisms-for-incomplete-stabilizers).

## Checklist and integration statements

- [x] My Python and C++ codes follow the coding and commenting styles of this project as indicated by existing files. Specifically, the changes conform to given `black`, `docformatter` and `pylint` configurations. 
- [x] I have performed a self-review of these changes, checked my code (including for [codefactor](https://www.codefactor.io/repository/github/xanaduai/flamingpy/branches) compliance), and corrected misspellings to the best of my capacity. I have synced this branch with others as required.
- [x] I have added context for corresponding changes in documentation and [`README.md`](README.md) as needed.
- [x] I have added new workflow CI tests for corresponding changes, ensuring codecoverage 95% or better, and these pass locally for me.
- [x] I have updated [`CHANGELOG.md`](.github/CHANGELOG.md) following the template. I recognize that the developers may revisit [`CHANGELOG.md`](.github/CHANGELOG.md) and the versioning, and create a Special Release including my changes.

### CODE REVIEW

**Comment by nariman87** [comment]
This needs more discussion to give users the option to swap axes and observe possible bugs.

**Comment by nariman87** [comment]
> Talking with @ilan-tz and @soosub after our meeting, we think that it would be nice to have the draw function work even when there are no bit values defined in the cluster state. For example, a solution could be drawing grey boxes rather than green or red boxes whenever the weights have not been assigned. It would be nice if this enhancement is included in this PR!

Thanks for the new changes @sduquemesa, looking fantastic.

@BestQuark, based on your comment above, are you happy to review this PR as the next step?

**Review by BestQuark — COMMENTED** [review]
( [see review here]((https://github.com/XanaduAI/flamingpy/pull/35#issuecomment-1152625329) )

**Review by BestQuark — APPROVED** [review]
Great PR! everything looks good to me. 

(Before mergin, can we quickly change line 374 of `viz.py` from 
`ax.scatter(x, y, z, c=color)` to `ax.scatter(x, y, z, c=color, s=4)`? I think this would make the plots a bit sharper).

File: flamingpy/utils/viz.py [hunk 1]
```diff
@@ -696,6 +701,51 @@ def syndrome_plot(code, ec, index_dict=None, drawing_opts=None):
...
+
+    # Find the size of the cube
+    gap = 0.15  # gap defines the distance between adjacent cubes
+    cube_origin = np.array([xmin, ymin, zmin]) + gap
+    max_arr = np.array([xmax, ymax, zmax])
+    size = np.abs(max_arr - cube_origin) - 2 * gap
+
+    # if there is any incomplete stabilizer, halve its size on
+    # the missing direction
+    is_incomplete_stab = len(cube_coords) != len(cube.physical)
```
Comment (ilan-tz) [inline flamingpy/utils/viz.py:726]:
```suggestion
    is_incomplete_stab = len(cube_coords) != 6
```

File: flamingpy/utils/viz.py [hunk 2]
```diff
@@ -696,6 +701,51 @@ def syndrome_plot(code, ec, index_dict=None, drawing_opts=None):
...
+    if is_incomplete_stab:
+        # to do so the cube is tranlated to the origin of coordinates
```
Comment (ilan-tz) [inline flamingpy/utils/viz.py:728]:
```suggestion
        # Translate the cube to the origin
```

File: flamingpy/utils/viz.py [hunk 3]
```diff
@@ -696,6 +701,51 @@ def syndrome_plot(code, ec, index_dict=None, drawing_opts=None):
...
+        centered_cube = cube_coords - np.tile(midpoint, (cube_coords.shape[0], 1))
+        # then stabilizers are paired (meaning 2 stabilizers) by finding the number
```
Comment (ilan-tz) [inline flamingpy/utils/viz.py:730]:
```suggestion
        #  Pair the stabilizers by finding the number
```

**Review by ilan-tz — CHANGES_REQUESTED** [review]
Really nice visual change here -- looks superb! Just to confirm: the issue with the x, y, z order in the stabilizer placement was fixed since our discussion, right? 

Unrelated to this PR, I think there is still an issue in some places with the x, y, z order -- e.g. the matching plot:

Figure_17

Could you please sweep through the viz file to see wherever else it may still be a problem?

File: flamingpy/codes/stabilizer.py [hunk 4]
```diff
@@ -48,10 +48,14 @@ def parity(self):
...
+            int, None: the total parity or ``None`` if bit values are not define
         """
-        bit_vals = [self.egraph.nodes[node]["bit_val"] for node in self.egraph]
-        par = int(np.sum(bit_vals) % 2)
+        bit_vals = [
+            self.egraph.nodes[node]["bit_val"]
+            for node in self.egraph
+            if "bit_val" in self.egraph.nodes[node]
+        ]
+        par = int(np.sum(bit_vals) % 2) if len(bit_vals) > 0 else None
```
[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/stabilizer.py:58]: Perfect, great suggestion! I have implemented it with a slight change.

[the remark this answers is not in the document]
Comment (sduquemesa) [reply flamingpy/codes/stabilizer.py]: Actually, your suggestion worked better than mine 🤭

**Comment by sduquemesa** [comment]
Thanks @ilan-tz and @BestQuark!

> Just to confirm: the issue with the x, y, z order in the stabilizer placement was fixed since our discussion, right?
> Unrelated to this PR, I think there is still an issue in some places with the x, y, z order
> Could you please sweep through the viz file to see wherever else it may still be a problem?

I reviewed the file and changed/commented where necessary. Please check if the issues you encountered were fixed.

image

> Before merging, can we quickly change line 374 of viz.py from `ax.scatter(x, y, z, c=color)` to `ax.scatter(x, y, z, c=color, s=4)` (or some small number to reduce the size of the egraph nodes)? I think this would make the plots a bit sharper.

Done! See the size of the nodes in the image above 🎱

**Comment by ilan-tz** [comment]
> Thanks @ilan-tz and @BestQuark!
> 
> > Just to confirm: the issue with the x, y, z order in the stabilizer placement was fixed since our discussion, right?
> > Unrelated to this PR, I think there is still an issue in some places with the x, y, z order
> > Could you please sweep through the viz file to see wherever else it may still be a problem?
> 
> I reviewed the file and changed/commented where necessary. Please check if the issues you encountered were fixed.
> 
> image
> 
> > Before merging, can we quickly change line 374 of viz.py from `ax.scatter(x, y, z, c=color)` to `ax.scatter(x, y, z, c=color, s=4)` (or some small number to reduce the size of the egraph nodes)? I think this would make the plots a bit sharper.
> 
> Done! See the size of the nodes in the image above 🎱

Looks nice! There is a small thing left to address:

- Have a look at the matching on the top-right (purple line). The left endpoint looks good, because it aligns with the middle of the stabilizer. However, the right endpoint sticks out, because it is being told to scatter at the midpoint of a complete stabilizer. Instead, it might look better if it terminates at the midpoint of an incomplete stabilizer (modification to line 826 of `viz.py`, or to the methods of the `Stabilizer` class (modify the `midpoint `method or add a a new method).
- The matching on the bottom left has the same issue. But since it starts at a boundary point, if we implement the above, the matching will go "diagonally" into the new midpoint. This might be okay! We will have to see how it looks.

**Comment by sduquemesa** [comment]
Hey @ilan-tz, this is how it looks like when I apply the (not yet committed) changes

image

What do you think? I feel it looks good between stabilizers (upper line) but a bit weird for node-to-stabilizer (lower diagonal line).

**Comment by sduquemesa** [comment]
After some more tweaking, this is how the end result looks like

image

The logic for drawing lines is the following

* If points belong to stabilizers, draw a line between the center of the stabilizers. 
* If the a point belongs to a stabilizer and another point is a node in the graph (this is the case for errors on the boundaries) then a line is drawn going from the midpoint of the stabilizer to the midpoint of the face on that boundary (and marked with an X on the plot), and another dotted line is drawn going from the midpoint of the face to the qubit where the error occurred.

**Comment by ilan-tz** [comment]
> After some more tweaking, this is how the end result looks like
> 
> [...]
> 
> The logic for drawing lines is the following
> 
> * If points belong to stabilizers, draw a line between the center of the stabilizers.
> * If the a point belongs to a stabilizer and another point is a node in the graph (this is the case for errors on the boundaries) then a line is drawn going from the midpoint of the stabilizer to the midpoint of the face on that boundary (and marked with an X on the plot), and another dotted line is drawn going from the midpoint of the face to the qubit where the error occurred.

Thanks @sduquemesa ! Looks nice, but when you have a moment, I think the following would be good to tweak:

- For the part ".. a line is drawn going from the midpoint of the stabilizer to the midpoint of the face on that boundary" there is some ambiguity. First is "that boundary" should never correspond to a plane along _z_ since we the _x_/_y_ directions correspond to the 2D code boundaries. Second -- let us say you have a node on the corner. That could correspond to the _x_ or _y_ boundary of the lattice. This can be decided based on `SurfaceCode.ec` -- "primal" corresponds to _x_ and "dual" to _y_.
- I am seeing the following issue, where the X is not on the boundary but in the middle:

image

**Pull request closed without merging**